File: VB/ColorSelectorControl.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.ColorPick.Picker
Imports DevExpress.XtraEditors.Popup
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraTab

Namespace DXSample
    Partial Public Class ColorSelectorControl
        Inherits XtraUserControl


        Private selectedColor_Renamed As Color

        Public Sub New()
            Me.selectedColor_Renamed = Color.Black
            InitializeComponent()
        End Sub
        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            InitControls()
        End Sub

        Public Property SelectedColor() As Color
            Get
                Return selectedColor_Renamed
            End Get
            Set(ByVal value As Color)
                Dim old As Color = value
                If SelectedColor = value Then
                    Return
                End If
                selectedColor_Renamed = value
                OnSelectedColorChanged(old, SelectedColor)
            End Set
        End Property
        Protected Overrides Sub OnMouseWheel(ByVal e As MouseEventArgs)
            MyBase.OnMouseWheel(e)
            Dim current As InnerColorPickControlBase = GetCurrentSelector()
            If current IsNot Nothing Then
                current.DoMouseWheel(e)
            End If
        End Sub

        Protected Function GetCurrentSelector() As InnerColorPickControlBase
            Dim page As XtraTabPage = Me.xtraTabControl.SelectedTabPage
            If page Is Nothing OrElse page.Controls.Count = 0 Then
                Return Nothing
            End If
            Return TryCast(page.Controls(0), InnerColorPickControlBase)
        End Function
        Private Sub OnSelectedColorChanged(ByVal oldColor As Color, ByVal newColor As Color)
            RaiseEvent SelectedColorChanged(Me, New InnerColorPickControlSelectedColorChangedEventArgs(New ColorItem(oldColor), New ColorItem(newColor)))
        End Sub

        Private Sub InitControls()
            InitializeSystemColors()
            InitializeWebColors()
            SubscribeEvents()
        End Sub

        Private Sub InitializeSystemColors()
            Me.systemColorListControl.Colors.AddColorRange(ColorListBoxViewInfo.SystemColors)
        End Sub

        Private Sub InitializeWebColors()
            Me.webColorListControl.Colors.AddColorRange(ColorListBoxViewInfo.WebColors)
        End Sub
        Private Sub SubscribeEvents()
            AddHandler Me.colorPickControl.SelectedColorChanged, AddressOf OnSelectedColorChanged
            AddHandler Me.colorPickControl.MoreButtonClick, AddressOf OnMoreButtonClick
            AddHandler Me.systemColorListControl.SelectedColorChanged, AddressOf OnSelectedColorChanged
            AddHandler Me.webColorListControl.SelectedColorChanged, AddressOf OnSelectedColorChanged
        End Sub

        Public Event MoreButtonClick As EventHandler

        Private Sub OnMoreButtonClick(ByVal sender As Object, ByVal e As EventArgs)
            RaiseEvent MoreButtonClick(Me, EventArgs.Empty)
            DoShowColorDialog()
        End Sub


        Private properties_Renamed As RepositoryItemColorPickEdit = Nothing
        Private ReadOnly Property Properties() As RepositoryItemColorPickEdit
            Get
                If properties_Renamed Is Nothing Then
                    properties_Renamed = New RepositoryItemColorPickEdit()
                End If
                Return properties_Renamed
            End Get
        End Property

        Protected Overridable Sub DoShowColorDialog()
            Using frm As New FrmColorPicker(Properties)
                frm.StartPosition = FormStartPosition.CenterScreen
                frm.SelectedColor = Me.selectedColor_Renamed
                If frm.ShowDialog(GetOwner()) = DialogResult.OK Then
                    Dim color As Color = frm.SelectedColor
                    DoSaveColor(color)
                End If
            End Using
        End Sub
        Protected Overridable Function GetOwner() As IWin32Window
            Return Nothing
        End Function
        Protected Overridable Sub DoSaveColor(ByVal color As Color)
            Properties.RecentColors.InsertColor(color, 0)
            Me.colorPickControl.RecentColors.Insert(0, New ColorItem(color))
            SelectedColor = color
        End Sub

        Public Function CalcBestSize() As Size
            Dim sz As New Size(Width, CalcBestHeight())
            Return Me.xtraTabControl.CalcSizeByPageClient(sz)
        End Function

        Private Const BorderSize As Integer = 2
        Protected Function CalcBestHeight() As Integer
            Return Me.colorPickControl.CalcBestHeight(10) + 2 * BorderSize
        End Function
        Private Sub OnSelectedColorChanged(ByVal sender As Object, ByVal e As InnerColorPickControlSelectedColorChangedEventArgs)
            SelectedColor = e.NewColor
        End Sub
        Public Event SelectedColorChanged As InnerColorPickControlSelectedColorChangedEventHandler
    End Class
End Namespace

File: CS/ColorSelectorControl.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.ColorPick.Picker;
using DevExpress.XtraEditors.Popup;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraTab;

namespace DXSample {
    public partial class ColorSelectorControl : XtraUserControl {
        Color selectedColor;
        
        public ColorSelectorControl() {
            this.selectedColor = Color.Black;
            InitializeComponent();
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            InitControls();
        }

        public Color SelectedColor {
            get { return selectedColor; }
            set {
                Color old = value;
                if(SelectedColor == value)
                    return;
                selectedColor = value;
                OnSelectedColorChanged(old, SelectedColor);
            }
        }
        protected override void OnMouseWheel(MouseEventArgs e) {
            base.OnMouseWheel(e);
            InnerColorPickControlBase current = GetCurrentSelector();
            if(current != null) current.DoMouseWheel(e);
        }

        protected InnerColorPickControlBase GetCurrentSelector() {
            XtraTabPage page = this.xtraTabControl.SelectedTabPage;
            if(page == null || page.Controls.Count == 0) return null;
            return page.Controls[0] as InnerColorPickControlBase;
        }
        void OnSelectedColorChanged(Color oldColor, Color newColor) {
            if(SelectedColorChanged != null) SelectedColorChanged(this, new InnerColorPickControlSelectedColorChangedEventArgs(new ColorItem(oldColor), new ColorItem(newColor)));
        }

        void InitControls() {
            InitializeSystemColors();
            InitializeWebColors();
            SubscribeEvents();
        }

        void InitializeSystemColors() {
            this.systemColorListControl.Colors.AddColorRange(ColorListBoxViewInfo.SystemColors);
        }

        void InitializeWebColors() {
            this.webColorListControl.Colors.AddColorRange(ColorListBoxViewInfo.WebColors);
        }
        void SubscribeEvents() {
            this.colorPickControl.SelectedColorChanged += OnSelectedColorChanged;
            this.colorPickControl.MoreButtonClick += OnMoreButtonClick;
            this.systemColorListControl.SelectedColorChanged += OnSelectedColorChanged;
            this.webColorListControl.SelectedColorChanged += OnSelectedColorChanged;
        }

        public event EventHandler MoreButtonClick;
        
        void OnMoreButtonClick(object sender, EventArgs e) {
            if(MoreButtonClick != null) MoreButtonClick(this, EventArgs.Empty);
            DoShowColorDialog();
        }

        RepositoryItemColorPickEdit properties = null;
        RepositoryItemColorPickEdit Properties {
            get {
                if(properties == null) properties = new RepositoryItemColorPickEdit();
                return properties;
            }
        }

        protected virtual void DoShowColorDialog() {
            using(FrmColorPicker frm = new FrmColorPicker(Properties)) {
                frm.StartPosition = FormStartPosition.CenterScreen;
                frm.SelectedColor = this.selectedColor;
                if(frm.ShowDialog(GetOwner()) == DialogResult.OK) {
                    Color color = frm.SelectedColor;
                    DoSaveColor(color);
                }
            }
        }
        protected virtual IWin32Window GetOwner() {
            return null;
        }
        protected virtual void DoSaveColor(Color color) {
            Properties.RecentColors.InsertColor(color, 0);
            this.colorPickControl.RecentColors.Insert(0, new ColorItem(color));
            SelectedColor = color;
        }

        public Size CalcBestSize() {
            Size sz = new Size(Width, CalcBestHeight());
            return this.xtraTabControl.CalcSizeByPageClient(sz);
        }

        const int BorderSize = 2;
        protected int CalcBestHeight() {
            return this.colorPickControl.CalcBestHeight(10) + 2 * BorderSize;
        }
        void OnSelectedColorChanged(object sender, InnerColorPickControlSelectedColorChangedEventArgs e) {
            SelectedColor = e.NewColor;
        }
        public event InnerColorPickControlSelectedColorChangedEventHandler SelectedColorChanged;
    }
}
